redis/src/cmd.rs

```diff
@@ -458,7 +465,7 @@ impl Cmd {
         let rv = con.req_command(&self)?;
 
         let (cursor, batch) = if rv.looks_like_cursor() {
-            from_owned_redis_value::<(u64, Vec<T>)>(rv)?
+            from_owned_redis_value::<(u64, Vec<RedisResult<T>>)>(rv)?
```

Comment:
instead of this, keep `from_owned_redis_value::<(u64, Vec<T>)>` and invert the result - Ok(vec) -> vec of Ok, and Err(err) -> vec![err].

Reply:
I was basically trying to implement this suggestion: https://github.com/redis-rs/redis-rs/pull/1416#discussion_r1845546102

Reply:
nvm, I was looking at the wrong diff

Reply:
>  invert the result - Ok(vec) -> vec of Ok, and Err(err) -> vec![err].

I'm not entirely sure what this means. Also I think if we change this, then when any one of the elements of the vec fails to parse into the type `T` , it will return an error and the whole operation will fail.

Reply:
```rs


/// Represents a redis iterator.
pub struct Iter<'a, T: FromRedisValue> {
    batch: std::vec::IntoIter<Value>,
    cursor: u64,
    con: &'a mut (dyn ConnectionLike + 'a),
    cmd: Cmd,
    _phantom: PhantomData<T>,
}

impl<'a, T: FromRedisValue> Iterator for Iter<'a, T> {
    type Item = RedisResult<T>;

    #[inline]
    fn next(&mut self) -> Option<RedisResult<T>> {
        // we need to do this in a loop until we produce at least one item
        // or we find the actual end of the iteration.  This is necessary
        // because with filtering an iterator it is possible that a whole
        // chunk is not matching the pattern and thus yielding empty results.
        loop {
            if let Some(v) = self.batch.next() {
                return Some(T::from_owned_redis_value(v));
            };
            if self.cursor == 0 {
                return None;
            }

            let pcmd = self.cmd.get_packed_command_with_cursor(self.cursor)?;
            let rv = match self.con.req_packed_command(&pcmd) {
                Ok(v) => v,
                Err(e) => return Some(Err(e)),
            };

            let (cur, batch): (u64, Vec<Value>) = match from_owned_redis_value(rv) {
                Ok(v) => v,
                Err(e) => return Some(Err(e)),
            };

            self.cursor = cur;
            self.batch = batch.into_iter();
        }
    }
}

Reply:
Hey @nihohit , sorry for the delay. I'm not too familiar with rust iterators, and I tried to fight the compiler a bit to implement the suggestion. I'm not sure if I understood correctly, but what I did seems to undo the prior suggestion (https://github.com/redis-rs/redis-rs/pull/1416#discussion_r1845546102) and bring back the infinite loop

Reply:
i.e. once the invalid UTF-8 is returned, it keeps scanning the same position:
![image](https://github.com/user-attachments/assets/87424173-7d5e-486b-b849-f783af88b1be)

Reply:
Unless `Value` means something else in the context of the suggestion.

Reply:
and I don't think you've implemented my suggestion - you keep pre-converting the values, so your batch is still a `Vec<T>` instead of `Vec<Value>`. I appreciate that you made the effort, Rust isn't always easy, but changing a type from my suggestion should've been a warning that you're doing something wrong :).
see my branch - https://github.com/nihohit/redis-rs/tree/safe-iterator, where the tests that fail are those that you modified, and still need to be solved. `test_scanning_error` passes.

Reply:
That's the spirit :)

Reply:
I think I've done it correctly this time, though just for iter right now (not async iter)